### PR TITLE
Replace old formatter with new formatter

### LINKED ISSUE #1516: [Tracking] `swayfmt-v2` MVP
Tracking for the new formatter:

>The design of `rustfmt` is to leave code functionality in tact, but also do more than just move white space. For instance,
>long generic parameters can be cleaned up with a `where` clause addition and formatting of those parameters after it. I
>think that's a really great characteristic that keeps code uniform and even teaches users the standard practices.
>
>It uses an `AST` for most of the formatting, but for macros it uses tokens or otherwise leaves them untouched. Most macros aren't formatted, unless necessary like the `format!` macro since it takes in variables that result in a `String`.
>
>`rustfmt` can recognize when something won't benefit from being formatted and opts that portion of the code out of the formatting process. If code doesn't fit the conditions of what would need formatting, e.g. a user created diagram, then it will
>just leave it alone.
>
Please feel free to hop in and take on some of this if it interests you:

- husk of the formatter, config settings & `swayfmt.toml` #1425 #1803 @eureka-cpu 
- imports #2295 @eureka-cpu 
- literals @eureka-cpu 
- whitespace #1888 #1894 #1911 @eureka-cpu 
- expressions #2338 @eureka-cpu 
- collections @eureka-cpu 
- user defined @kayagokalp 
- items
  - abi #2043 #2044 @eureka-cpu 
  - const #2005 @kayagokalp 
  - enum #1873 @kayagokalp 
  - fn #2173  @eureka-cpu 
  - impl #2403 @eureka-cpu 
  - storage #2069 @kayagokalp 
  - struct #2053 @kayagokalp 
  - trait #2294 @eureka-cpu 
  - use #2295 @eureka-cpu 
  - comments ( blocker: #1517 ) @mitchmindtree & @kayagokalp (#2229, #2311)
  - `break` and `continue` https://github.com/FuelLabs/sway/issues/2118 (pending https://github.com/FuelLabs/sway/pull/2112) @eureka-cpu 
- attributes #2012 #2061 @eureka-cpu 
- Contract call arguments: https://github.com/FuelLabs/sway/issues/992 #2338 @eureka-cpu 
- long conditionals #2597 @eureka-cpu 

`swayfmt`'s goal is to be somewhat transparent. It should format Sway code to our standards, and if the user so decides it, they can add a `swayfmt.toml` to their root dir and it will assume only the changes present, and leave the rest to default.

Once completed add to CI documentor: #1806

### PR DESCRIPTION
**Pull request by eureka-cpu**
Closes #1516 

- [x] Rename and replace old fmt with new fmt
- [x] Fix any conflicts that arise from this
- [ ] #2691

### CODE REVIEW

**Comment by mitchmindtree** [comment]
This is legendary @eureka-cpu! :zap: :rocket: :100:

File: examples/hashing/src/main.sw [hunk 1]
```diff
@@ -73,36 +92,57 @@ fn main() {
     log(sha_hashed_enum);
     log(sha_hashed_struct);
 
+
```
Comment (mohammadfawaz) [inline examples/hashing/src/main.sw:95]: What is the reason for the extract newlines here and in other places?

Comment (mohammadfawaz) [reply examples/hashing/src/main.sw:95]: and does this mean that when I rerun the formatter another new line will be added?

Comment (mitchmindtree) [reply examples/hashing/src/main.sw:95]: @mohammadfawaz I just noticed this issue https://github.com/FuelLabs/sway/issues/2691 - @eureka-cpu possibly related?

**Comment by emilyaherbert** [comment]
Omg it's happening

File: examples/hashing/src/main.sw (hunk 1 above)
Comment (eureka-cpu) [reply examples/hashing/src/main.sw:95]: @mohammadfawaz & @mitchmindtree yes, there's an issue currently where newlines will continue to get added if there is a comment with a newline in its span. Doesn't seem to happen for comments that don't have an extra newline or in places without comments.

File: sway-lsp/src/sway_config.rs [hunk 2]
```diff
@@ -34,11 +42,24 @@ impl SwayConfig {
 // note `FormattingOptions` and `SwayConfig` may be similar at this moment,
 // but they are not the same thing, `SwayConfig` contains all the data related to the LanguageServer
 // while `FormattingOptions` is only the part that is necessary for 'formating'
-impl From<SwayConfig> for FormattingOptions {
+impl From<SwayConfig> for Formatter {
```
Comment (mitchmindtree) [inline sway-lsp/src/sway_config.rs:45]: Perhaps a closer mapping to the old behaviour would be:
```suggestion
impl From<SwayConfig> for Config {
```
where `Config` is the formatter's `Config` type?

Eventually we might want to simply remove `SwayConfig` in favour of using the formatter's `Config` directly, but we can discuss this with @JoshuaBatty once he's back in the office. For now, I think you're on the right track with simply using the defaults where necessary :+1:

**Comment by mitchmindtree** [comment]
@eureka-cpu I've just merged in #2698 (which should have @kayagokalp's fixes for the newline issues) into your branch here!

I was expecting this to resolve these `sway-fmt-v2` conflicts as there were no conflicts that arose locally when merging in `master`, but it seems they remain on this PR :thinking: Maybe GitHub has some approach to merging that differs from a simple local `git merge`?

If you have trouble resolving these conflicts within this PR, feel free to open up a new one as it's more important we land this sooner than later :)

**Comment by eureka-cpu** [comment]
> @eureka-cpu I've just merged in #2698 (which should have @kayagokalp's fixes for the newline issues) into your branch here!
> 
> I was expecting this to resolve these `sway-fmt-v2` conflicts as there were no conflicts that arose locally when merging in `master`, but it seems they remain on this PR thinking Maybe GitHub has some approach to merging that differs from a simple local `git merge`?
> 
> If you have trouble resolving these conflicts within this PR, feel free to open up a new one as it's more important we land this sooner than later :)

When I pull the branch it doesn't show any conflicts 🤔

**Comment by eureka-cpu** [comment]
Merging master did the trick oddly enough

File: sway-lsp/src/sway_config.rs (hunk 2 above)
Comment (eureka-cpu) [reply sway-lsp/src/sway_config.rs:45]: Checked in on this and while converting to `Config` may be closer to the previous code, it doesn't seem that we're using it anywhere else other than for the end result of constructing a `Formatter` and calling `.format()`. In other words we would be constructing a `Config` and then construct a `Formatter` and then replace the formatter's config with said config.

File: examples/hashing/src/main.sw [hunk 3]
```diff
@@ -35,30 +34,50 @@ fn main() {
     let sha_hashed_u32 = sha256(~u32::max());
     let sha_hashed_u64 = sha256(~u64::max());
 
+
```
Comment (Braqzen) [inline examples/hashing/src/main.sw:37]: Additional new lines? This happens in multiple files

File: examples/match_statements/src/main.sw [hunk 4]
```diff
@@ -1,12 +1,11 @@
...
+fn foo() {}
     // do something
-}
-fn bar() {
-    // do something
-}
+fn bar() {}
 
+
+    // do something
```
Comment (Braqzen) [inline examples/match_statements/src/main.sw:8]: Comments are broken here

File: examples/structs/src/main.sw [hunk 5]
```diff
@@ -40,73 +45,52 @@ fn shorthand_instantiation() -> Foo {
...
-    let Point {
-        x, y
-    }
-    = point1;
-
-    let point2 = Point {
-        x: 1,
-        y: 1,
-    };
+    let Point{x, y} = point1;
```
Comment (Braqzen) [inline examples/structs/src/main.sw:64]: This should be `let Point {x, y}` or imo `let Point { x, y }` is better

File: sway-lsp/src/server.rs [hunk 6]
```diff
@@ -79,7 +79,7 @@ fn capabilities() -> ServerCapabilities {
             trigger_characters: None,
             ..Default::default()
         }),
-        document_formatting_provider: Some(OneOf::Left(true)),
+        // document_formatting_provider: Some(OneOf::Left(true)),
```
Comment (Braqzen) [inline sway-lsp/src/server.rs:82]: Commented out line?

File: examples/hashing/src/main.sw (hunk 3 above)
Comment (eureka-cpu) [reply examples/hashing/src/main.sw:37]: Ah this is from the previous formatting I did, I'll go in and fix that rq

File: examples/match_statements/src/main.sw (hunk 4 above)
Comment (eureka-cpu) [reply examples/match_statements/src/main.sw:8]: I'll open an issue to address it, this happens when placing comments inside an empty block

File: examples/hashing/src/main.sw (hunk 3 above)
Comment (eureka-cpu) [reply examples/hashing/src/main.sw:37]: As it turns out this is actually a side effect of the previous bug fix for preceding newlines on comments. I'll reopen the issue.

Comment (eureka-cpu) [reply examples/hashing/src/main.sw:37]: #2691

Comment (eureka-cpu) [reply examples/hashing/src/main.sw]: I added a bandaid fix for now, and will continue looking into it further

**Review by Braqzen — APPROVED** [review]

**Review by mohammadfawaz — APPROVED** [review]

**Pull request merged**